lib/AST/ASTMangler.cpp

```diff
@@ -2287,10 +2317,32 @@ void ASTMangler::appendFunctionType(AnyFunctionType *fn, bool isAutoClosure,
     return appendOperator("c");
 
   case AnyFunctionType::Representation::CFunctionPointer:
+    if (mangleClangType) {
+      appendOperator("XzC");
+      return appendClangType(fn);
+    }
     return appendOperator("XC");
   }
 }
 
+template <typename FnType>
+void ASTMangler::appendClangType(FnType *fn, llvm::raw_svector_ostream &out) {
+  auto clangType = fn->getClangTypeInfo().getType();
+  SmallString<64> scratch;
+  llvm::raw_svector_ostream scratchOS(scratch);
+  clang::ASTContext &clangCtx = fn->getASTContext()
+      .getClangModuleLoader()
+      ->getClangASTContext();
+  std::unique_ptr<clang::ItaniumMangleContext> mangler{
+      clang::ItaniumMangleContext::create(clangCtx, clangCtx.getDiagnostics())};
```

Comment:
Do we always want `ItaniumMangleContext` for C++ types? On windows shouldn't we use `MSVC`?

Reply:
Actually, maybe Windows Itanium will be sufficient, never mind.

Reply:
Yeah, this doesn't affect interoperation, and we're going to appreciate using a unified mangling for both testing and (if we ever need it) dynamic-demangling purposes.

MSVC's mangling is also missing a fair amount of the complexity of the C type system supported by Clang.  It's also, in my opinion, pretty awful to try to read because the density of `@`s.

Reply:
> Yeah, this doesn't affect interoperation, and we're going to appreciate using a unified mangling for both testing and (if we ever need it) dynamic-demangling purposes.

Fair enough. Makes sense.